src/QuadForm/indefiniteLLL.jl

```diff
@@ -0,0 +1,316 @@
+
+@doc Markdown.doc"""
+    lll_gram_indef(G::MatElem{fmpz}, base::Bool = false) 
+                                        -> Tuple{MatElem{fmpz}, MatElem{fmpz}, MatElem{fmpz}}
+
+Given a Gram matrix `G` of an indefinite quadratic lattice with $det(G) \neq 0$,
+if an isotropic vector is found, return `G`, `I` and `sol` where `I` is the 
+identity-matrix and `sol` is the isotropic vector. Otherwise return a LLL-reduction 
+of `G`, the transformation matrix `U` and fmpz[].
+"""
+function lll_gram_indef(G::MatElem{fmpz}; base::Bool = false)
+  n = ncols(G)
+  M = identity_matrix(ZZ,n)
+  QD = G
+
+  # GSO breaks off if one of the basis vectors is isotropic
+  for i = 1:n-1
+    if QD[i,i] == 0
+      return _quad_form_solve_triv(G; base = base)
+    end
+
+    M1 = identity_matrix(QQ,n)
+    for j = 1:n
+      M1[i,j] = - QD[i,j]//QD[i,i]
+    end
+
+    M1[i,i] = 1
+    M = M*M1
+    QD = transpose(M1)*QD*M1
+  end
+
+  M = inv(M)
+  QD = transpose(M)*map_entries(abs, QD)*M
+  QD = QD*denominator(QD)
+  QD_cont = divexact(QD,content(QD))
+  QD_cont = change_base_ring(ZZ,QD_cont)
+  rank_QD = rank(QD_cont)
+
+  S = transpose(lll_gram_with_transform(QD_cont)[2])
+  S = S[:,ncols(S)-rank_QD+1:ncols(S)]
+
+  if ncols(S) < n
+    S = _complete_to_basis(S)
+  end
+
+  red = _quad_form_solve_triv(transpose(S)*G*S; base = base)
+  r1 = red[1]
+  r2 = red[2]
+  r3 = red[3]
+
+  if r3 != fmpz[] && base == false
+    r3 = S * r3
+    return r1, r2,r3
+  end
+
+  r2 = S*red[2]
+  
+  if r3 != fmpz[] && base == true
+    r3 = S*r3
+    return r1,r2,r3
+  end
+
+  return r1,r2,r3
+end
+
+
+@doc Markdown.doc"""
+    lll_gram_indefgoon(G::MatElem{fmpz}, check::Bool = false) 
+                                              -> Tuple{MatElem{fmpz}, MatElem{fmpz}}
+
+Perform the LLL-reduction of the Gram matrix `G` of an indefinite quadratic 
+lattice which goes on even if an isotropic vector is found. 
+If `check == true` the function checks  whether `G` is a symmetric, $det(G) \neq 0$ 
+and the Gram matrix of a non-degenerate, indefinite Z-lattice. 
+"""
+function lll_gram_indefgoon(G::MatElem{fmpz}; check::Bool = false)
+
+  if(check == true)
+    if(issymmetric(G) == false || det(G) == 0 || _isindefinite(change_base_ring(QQ,G)) == false)
+      error("Input should be a Gram matrix of a non-degenerate indefinite quadratic lattice.")
+    end
+  end
+
+  red = lll_gram_indef(G; base = true)
+  
+  #If no isotropic vector is found
+  if red[3] == fmpz[]
+    return red[1] , red[2]
+  end
+
+  U1 = red[2]
+  G2 = red[1]
+  U2 = _mathnf(G2[1,:])[2]
+  G3 = transpose(U2)*G2*U2
+  
+  #The first line of the matrix G3 only contains 0, except some 'g' on the right, where g² | det G.
+  n = ncols(G)
+  U3 = identity_matrix(ZZ,n)
+  U3[1,n] = round(- G3[n,n]//2*1//G3[1,n])
+  G4 = transpose(U3)*G3*U3
+
+  #The coeff G4[n,n] is reduced modulo 2g
+  U = G4[[1,n],[1,n]]
+
+  if (n == 2)
+    V = zero_matrix(ZZ,n,1)
+  else
+    V = _vecextract(G4, [1,n] , 1<<(n-1)-2)
```

Comment:
Here if I am not mistaking, you are taking the first and $n$-th row of `G4` and only the components (columns) from 2 to $n-1$ (indeed, the binary expansion of $1<<(n-1)-2 = 2^(n-1)-2$ is $1...110$ with $n-2$ 1s, so at the end you extract the columns in the range `2:(n-1)`. Your function is right, but it is the only instance where this `_vecextract` is used and I guess D. Simon used this notation as a shortcut, where you could use, here 
```
V = G4[[1,n], 2:(n-1)]
```
which would explain why you have to treat the case $n = 2$ apart. I guess for anybody reading this it would be easier to have it in this way because it is more " julia-like" and less "gp-type" code. Then the function `_vecextract` won't be too necessary.